### PR TITLE
[Hardware][Ascend] Add MindIE-SD diffusion compile backend

### PR DESCRIPTION
**Pull request by xiao6pinggai**
## Purpose

This PR adds MindIE-SD as a diffusion `torch.compile` backend for Ascend NPU and supports optional ACLGraph capture and replay on top of MindIE-SD pattern fusion. CUDA continues to use the existing Inductor path.

The main changes are:

- Add `diffusion_compile_backend` with `auto`, `inductor`, and `mindiesd` options. On Ascend, `auto` selects MindIE-SD when it is available and the requested compilation configuration is supported; CUDA behavior remains unchanged.
- Add `diffusion_compile_aclgraph` as an independent option. MindIE-SD pattern fusion works without ACLGraph, while `--diffusion-compile-aclgraph` explicitly enables graph capture and replay.
- Preserve the existing `regional` and `full` compilation granularities. Full-graph enforcement is applied only when full granularity and ACLGraph are enabled together.
- Keep `--enforce-eager` as the common eager-mode switch and bypass diffusion backend resolution when it is set.
- Resolve the diffusion backend before model loading so MindIE-SD runtime components and custom operators are registered before model initialization.
- Raise an actionable error when `mindiesd` is selected explicitly but is unavailable or incompatible. In `auto` mode, preserve the compatible eager fallback.
- Select model-specific residual-gate patterns by the loaded model class: `QwenImage*` uses the Qwen pattern, `Wan*` uses the Wan pattern, and unrelated models enable neither pattern.
- Move QwenImage all-valid attention-mask canonicalization outside the denoising loop to avoid a data-dependent graph break in each compiled transformer step.
- Resolve the MindIE-SD FlashAttention operator type during initialization instead of tracing environment-variable access inside the compiled graph.
- Propagate the new configuration through the CLI, Python API, stage configuration, worker, and platform interface, with documentation and test coverage.

MindIE-SD remains an optional Ascend dependency and is not added as a dependency for CUDA or other platforms. The QwenImage Transformer equations are unchanged.

### Companion MindIE-SD change

This vLLM-Omni change depends on MindIE-SD commit `44be6cb5e6ec3f0272373b8a757eee9fa1ee7db4`, which provides the QwenImage residual-gate pattern and the MindIE-SD ACLGraph/FlashAttention support used by this integration.

The corresponding vLLM-Omni implementation commit is `f058a1a33b6bfb6a9168567167af04e7c6596960`.

### Environment

- Hardware: one Ascend 910B (A3) NPU
- Model: Qwen-Image
- Precision: BF16
- Input: 512x512, fixed prompt, seed, CFG scale, batch size, and execution configuration
- Request pattern: concurrency 1; four warm-up requests followed by one measured request
- vLLM-Ascend commit: `09ca138b352fbcd7398a633d7962044bc84cce1e`
- MindIE-SD commit: `44be6cb5e6ec3f0272373b8a757eee9fa1ee7db4`
- Dependencies: `torch==2.10.0`, `torch_npu==2.10.0.post4`, `triton==3.5.0`, `triton-ascend==3.2.2`

**vLLM Version:** `84030bbe3d74d99bad477a3d2e37a973ccd8865c` (`v0.28.1rc0-676-g84030bbe3d`)

**vLLM-Omni Commit:** `f058a1a33b6bfb6a9168567167af04e7c6596960`

**vLLM-Omni Base Commit:** `92715f3f7b67d7835c08002309402ba08b7b237e`

### Server configurations

`xx/Qwen-Image/` represents the Qwen-Image model directory. vLLM, vLLM-Ascend, vLLM-Omni, and MindIE-SD are installed in the test environment.

#### Eager

```bash
ASCEND_RT_VISIBLE_DEVICES=8 \
vllm serve \
  xx/Qwen-Image/ \
  --host 127.0.0.1 \
  --port 33547 \
  --omni \
  --trust-remote-code \
  --disable-log-stats \
  --tensor-parallel-size 1 \
  --step-execution \
  --max-num-seqs 8 \
  --enforce-eager \
  --cache-backend none
```

#### MindIE-SD pattern fusion

```bash
ASCEND_RT_VISIBLE_DEVICES=8 \
vllm serve \
  xx/Qwen-Image/ \
  --host 127.0.0.1 \
  --port 33547 \
  --omni \
  --trust-remote-code \
  --disable-log-stats \
  --tensor-parallel-size 1 \
  --step-execution \
  --max-num-seqs 8 \
  --diffusion-compile-backend mindiesd \
  --diffusion-compile-granularity full \
  --no-diffusion-compile-dynamic \
  --cache-backend none
```

#### MindIE-SD pattern fusion with ACLGraph

```bash
ASCEND_RT_VISIBLE_DEVICES=8 \
vllm serve \
  xx/Qwen-Image/ \
  --host 127.0.0.1 \
  --port 33547 \
  --omni \
  --trust-remote-code \
  --disable-log-stats \
  --tensor-parallel-size 1 \
  --step-execution \
  --max-num-seqs 8 \
  --diffusion-compile-backend mindiesd \
  --diffusion-compile-granularity full \
  --no-diffusion-compile-dynamic \
  --diffusion-compile-aclgraph \
  --cache-backend none
```

The Qwen residual-gate comparison is controlled by MindIE-SD `FusionPatterns`. This PR does not add a model-specific environment variable or CLI option.

### Request command

The following command sends four sequential warm-up requests and one measured request using the same payload:

```bash
REQUEST_BODY='{
  "model": "xx/Qwen-Image/",
  "prompt": "Generate a realistic-style image: a red sports car parked by a calm lake.",
  "size": "512x512",
  "num_inference_steps": 50,
  "true_cfg_scale": 4.0,
  "seed": 42,
  "response_format": "b64_json",
  "output_format": "png"
}'

for _ in 1 2 3 4; do
  curl -sS http://127.0.0.1:33547/v1/images/generations \
    -H 'Content-Type: application/json' \
    -d "$REQUEST_BODY" \
    -o /dev/null
done

curl -sS http://127.0.0.1:33547/v1/images/generations \
  -H 'Content-Type: application/json' \
  -d "$REQUEST_BODY" \
  -o response.json \
  -w 'HTTP=%{http_code} total=%{time_total}s\n'
```

### Automated tests

```bash
pytest -q tests/platforms/test_diffusion_compile_backend.py
pytest -q tests/diffusion/test_compile.py tests/diffusion/test_diffusion_model_runner.py
pytest -q tests/config/test_omni_config.py tests/entrypoints/test_async_omni_diffusion_config.py
```

These tests cover configuration defaults and propagation, platform-specific backend selection, missing or incompatible MindIE-SD handling, ACLGraph propagation, lazy backend invocation, and QwenImage attention-mask preprocessing.

### End-to-end validation

- Qwen-Image service startup: passed
- Four warm-up requests followed by one measured 512x512 request: passed
- Eager, MindIE-SD pattern fusion, and MindIE-SD pattern fusion with ACLGraph: all produced valid images
- Non-Qwen model handling is protected by model-class-based pattern selection; Qwen and Wan residual-gate patterns are not enabled for unrelated model classes

### Performance and operator count

| Mode | 50-step latency | Improvement vs. eager | Device ops per DiT block | Peak memory |
| --- | ---: | ---: | ---: | ---: |
| Eager | 7912 ms | - | 102 | 56288 MiB |
| Pattern without Qwen residual gate | 7155 ms | 9.57% | 69 | 56284 MiB |
| Pattern with Qwen residual gate | 7196 ms | 9.05% | 65 | 56284 MiB |
| Pattern + ACLGraph without Qwen residual gate | 6974 ms | 11.86% | - | 56562 MiB |
| Pattern + ACLGraph with Qwen residual gate | 6962 ms | 12.01% | - | 56564 MiB |

MindIE-SD pattern fusion reduces the per-block device-op count from 102 to 69. The Qwen residual-gate pattern further reduces it to 65, while its standalone latency benefit is within the observed run-to-run variation. Pattern fusion with ACLGraph provides the largest measured 50-step latency improvement at 12.01%.

Peak memory remains effectively unchanged with pattern fusion alone (56288 MiB to 56284 MiB). Enabling ACLGraph increases peak memory slightly to 56562-56564 MiB, approximately 274-276 MiB (0.49%) above eager execution.

### Output consistency

All images use identical request parameters and are original 512x512 RGB PNG outputs without resizing or re-encoding.

The mean absolute RGB difference was calculated pixel-by-pixel across all three channels:

| Comparison | Mean absolute RGB difference | Result |
| --- | ---: | --- |
| Eager vs. pattern without Qwen residual gate | 0.5170 | Minor pixel-level difference |
| Eager vs. pattern with Qwen residual gate | 1.2189 | Minor pixel-level difference |
| Pattern with vs. without Qwen residual gate | 0.8718 | Minor pixel-level difference |
| Pattern vs. Pattern + ACLGraph, with Qwen residual gate | 0.0000 | Byte-identical |
| Pattern vs. Pattern + ACLGraph, without Qwen residual gate | 0.0000 | Byte-identical |

Eager and pattern-fusion outputs have small pixel-level differences. With the residual-gate configuration held constant, ACLGraph introduces no additional output difference.

Note: Practical tests show the "Regional" mode in ACLGraph has lower speed and heavily degraded output image quality. Further optimization is needed.

### CODE REVIEW

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 1]
```diff
@@ -729,6 +736,12 @@ def _prepare_generation_context(
             negative_prompt_embeds = None
             negative_prompt_embeds_mask = None
 
+        # This reduction is request-level work. Keeping it outside transformer
+        # forward prevents one data-dependent Dynamo graph break per denoise
+        # invocation while preserving real padding masks.
+        prompt_embeds_mask = canonicalize_qwen_image_attention_mask(prompt_embeds_mask)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py:742]: Keep request masks until batching is done, then simplify the combined mask.

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py [hunk 2]
```diff
@@ -1209,8 +1209,9 @@ def forward(
                 self.parallel_config.sequence_parallel_size,
             )
 
-        if encoder_hidden_states_mask is not None and encoder_hidden_states_mask.all():
-            encoder_hidden_states_mask = None
+        # All-valid masks are canonicalized once by QwenImagePipeline before
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:1212]: Handle all-valid masks in Edit, EditPlus, and Layered before dropping this fast path.

File: vllm_omni/platforms/npu/platform.py [hunk 3]
```diff
@@ -292,6 +318,88 @@ def supports_diffusion_dense_flash_attention(cls) -> bool:
...
+            if allow_eager_fallback:
+                logger.info("MindIE-SD is unavailable; running the diffusion pipeline eagerly.")
+                return None
+            raise RuntimeError(
+                "MindIE-SD diffusion compilation was requested, but mindiesd.compilation could not be imported. "
+                "Install a MindIE-SD build compatible with PyTorch, torch_npu and CANN, "
+                "or use enforce_eager=True."
+            ) from exc
+
+        incompatibility = None
```
Comment (lishunyang12) [inline vllm_omni/platforms/npu/platform.py:343]: Let's move the compatibility checks and CompilationConfig setup into separate helpers.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py (hunk 1 above)
Comment (xiao6pinggai) [reply vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: Fixed. Request-local masks now remain on `StepRequestState` until `InputBatch` pads and combines the requests. `QwenImagePipeline.denoise_step()` canonicalizes only the completed batch mask. A mixed-length regression test verifies that a two-token request batched with a four-token request keeps the expected padding mask, while an equal-length batch drops its all-valid mask.

File: vllm_omni/platforms/npu/platform.py (hunk 3 above)
Comment (xiao6pinggai) [reply vllm_omni/platforms/npu/platform.py]: Fixed. `_get_mindiesd_incompatibility()` now owns compatibility validation, and `_configure_mindiesd_compilation()` owns `CompilationConfig` and model-specific fusion-pattern setup. `get_diffusion_compile_backend()` is limited to backend selection, optional dependency handling, fallback policy, and backend construction.

File: vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py (hunk 2 above)
Comment (xiao6pinggai) [reply vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py:1251]: Fixed. All-valid mask canonicalization for the non-step path now lives in `QwenImageCFGParallelMixin.diffuse()`, which is shared by Qwen-Image, Qwen-Image Edit, Qwen-Image EditPlus, and Qwen-Image Layered. The transformer no longer performs the data-dependent reduction inside the compiled forward path. A parameterized regression test verifies that all four pipelines use the shared implementation.

**Comment by xiao6pinggai** [comment]
@vllm-omni-review-bot Please review.

**Comment by xiao6pinggai** [comment]
I reviewed the updated diff for backend selection and eager fallback behavior, model-specific residual-gate selection, ACLGraph configuration propagation, Qwen-Image mask batching, and CUDA/other-platform
  isolation. I also checked that request-level masks are not simplified before batching and that all four Qwen-Image pipeline variants share the same non-step canonicalization path.

 Validation performed without rebuilding the runtime:

  - `tests/diffusion/models/qwen_image/test_qwen_image_max_sequence_length.py`
  - `tests/platforms/test_diffusion_compile_backend.py`
  - Result: 65 passed
  - `git diff --check`: passed
